src/Runtime/Distributions/Automata/Automaton.cs

```diff
@@ -1011,45 +1030,49 @@ public void SetGroup(int group)
         /// <returns>The logarithm of the normalizer.</returns>
         /// <exception cref="InvalidOperationException">Thrown if the automaton cannot be normalized (i.e. if the normalizer is zero or positive infinity).</exception>
         /// <remarks>The only automaton which cannot be normalized, but has a finite normalizer, is zero.</remarks>
-        public double NormalizeValues()
+        public double NormalizeValues(out TThis result)
         {
-            double result;
-            if (!this.TryNormalizeValues(out result))
+            if (!TryNormalizeValues(out result, out double logNormalizer))
             {
                 throw new InvalidOperationException("This automaton cannot be normalized.");
             }
 
-            return result;
+            return logNormalizer;
         }
 
         /// <summary>
-        /// Attempts to normalize the automaton so that sum of its values on all possible sequences equals to one (if it is possible).
+        /// Attempts to normalize the automaton so that sum of its values on all possible sequences equals to one (if it is possible)
+        /// and returns the result in an out parameter.
+        /// If successful, produces a stochastic automaton,
+        /// i.e. an automaton, in which the sum of weights of all the outgoing transitions and the ending weight is 1 for every node.
         /// </summary>
+        /// <param name="normalizedAutomaton">Will be set to the normalized automaton if the normalization was successful, or to the current automaton otherwise.</param>
         /// <param name="logNormalizer">When the function returns, contains the logarithm of the normalizer.</param>
-        /// <returns><see langword="true"/> if the automaton was successfully normalized, <see langword="false"/> otherwise.</returns>
-        public bool TryNormalizeValues(out double logNormalizer)
+        /// <returns><see langword="true"/> if the normalization was successful, <see langword="false"/> otherwise.</returns>
+        public bool TryNormalizeValues(out TThis normalizedAutomaton, out double logNormalizer)
         {
-            logNormalizer = this.DoGetLogNormalizer(true);
+            logNormalizer = DoGetLogNormalizer(true, out normalizedAutomaton);
             return !double.IsInfinity(logNormalizer);
         }
 
         /// <summary>
         /// Attempts to normalize the automaton so that sum of its values on all possible sequences equals to one (if it is possible).
         /// </summary>
+        /// <param name="result">Will be set to the normalized automaton if the normalization was successful, or to the current automaton otherwise.</param>
         /// <returns><see langword="true"/> if the automaton was successfully normalized, <see langword="false"/> otherwise.</returns>>
-        public bool TryNormalizeValues()
+        public bool TryNormalizeValues(out TThis result)
```

Comment:
EnsureNormalized should use this overload

Reply:
EnsureNormalized operates on `TWeightFunction : IWeghtFunction<...>`. To keep `IWeightFunction<...>` simpler and not allow potential inconsistencies I've out only one overload of `TryNormalizeValues` (the one with both outputs) there.